atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/SingleBusyCellWorkflows.java

```diff
@@ -0,0 +1,110 @@
+    private SingleBusyCellWorkflows() {
+        // utility
+    }
+
+    @VisibleForTesting
+    static final WorkloadCell BUSY_CELL = ImmutableWorkloadCell.of(1, 1);
+
+    public static Workflow create(
+            InteractiveTransactionStore store,
+            SingleBusyCellWorkflowConfiguration configuration,
+            ListeningExecutorService readExecutor,
+            ListeningExecutorService writeExecutor) {
+        return () -> {
+            List<ListenableFuture<Optional<WitnessedTransaction>>> reads =
+                    scheduleReadsAndTouches(store, configuration, readExecutor);
+            List<ListenableFuture<Optional<WitnessedTransaction>>> writes =
+                    scheduleWrites(store, configuration, writeExecutor);
+
+            ListenableFuture<List<WitnessedTransaction>> witnessedTransactions = Futures.transform(
+                    Futures.allAsList(
+                            Stream.of(reads, writes).flatMap(Collection::stream).collect(Collectors.toList())),
+                    maybeWitnessedTransactions -> maybeWitnessedTransactions.stream()
+                            .flatMap(Optional::stream)
+                            .collect(Collectors.toList()),
+                    MoreExecutors.directExecutor());
+            ReadOnlyTransactionStore readOnlyTransactionStore = new ReadOnlyTransactionStore(store);
+            return ImmutableWorkflowHistory.builder()
+                    .addAllHistory(WitnessedTransactions.sortAndFilterTransactions(
+                            readOnlyTransactionStore, Futures.getUnchecked(witnessedTransactions)))
+                    .transactionStore(readOnlyTransactionStore)
+                    .build();
+        };
+    }
+
+    @NotNull
+    private static List<ListenableFuture<Optional<WitnessedTransaction>>> scheduleReadsAndTouches(
+            InteractiveTransactionStore store,
+            SingleBusyCellWorkflowConfiguration configuration,
+            ListeningExecutorService readExecutor) {
+        return IntStream.range(0, configuration.iterationCount() / 2)
+                .mapToObj(idx -> readExecutor.submit(() -> store.readWrite(txn -> {
+                    Optional<Integer> value =
+                            txn.read(configuration.tableConfiguration().tableName(), BUSY_CELL);
+                    txn.write(configuration.tableConfiguration().tableName(), BUSY_CELL, value.orElse(-1));
+                })))
+                .collect(Collectors.toList());
+    }
+
+    @NotNull
+    private static List<ListenableFuture<Optional<WitnessedTransaction>>> scheduleWrites(
+            InteractiveTransactionStore store,
+            SingleBusyCellWorkflowConfiguration configuration,
+            ListeningExecutorService writeExecutor) {
+        List<ListenableFuture<Optional<WitnessedTransaction>>> writes = new ArrayList<>();
+        for (int index = 0; index < configuration.iterationCount() / 4; index++) {
+            int stableIndex = index;
```

Comment:
Why not just use `IntStream`? Why do you need this `stableIndex`? Index itself should not be mutated within the loop?

Reply:
Ah yeah we should just use `IntStream`, I think this was a leftover I didn't clean up when refactoring.

Though note that if you want the for loop you **need** the `stableIndex`, because Java cares that the variable is final and `index` definitely is not - it doesn't matter whether it's mutated within the loop or not (or even if the variable is effectively final at the point of usage). For example, even something like this will be rejected

```
int i = 5;
writeExecutor.submit(() -> i); // Nope
i = 8;
```

Reply:
https://docs.oracle.com/javase/specs/jls/se10/html/jls-15.html#jls-15.27.2 has gory details if you like 😅

Reply:
(I suspect you're thinking of the special rule for enhanced for loops!)